### DISCUSSION TITLE
graph shows an unexpected app node (not part of applications list).

### DISCUSSION THREAD

**Discussion by saarwasserman** [kiali/kiali#4214] [Q&A] [answered]
kiali version: 1.34.1

I use 'app' as the application label. In the applications tab I see all the applications as expected.
However in the graph I see different app name.
When clicking the app link on graph side-bar it is not linked to any app so I get 404.

from which field the graph creates the app node?
How this value it is set?
is it related to the 'destination_canonical_service' field? I saw it's part of the prometheus query in the logs.

note: the app.kubernetes.io/name is also set but I defined the 'app' label to be used in kiali cr and it works as I can see the apps in applications tab.

**Comment by jshaughn** [comment, marked as answer]
The graph always determines app and version by using the `canonical_service` and `canonical_revision` attributes in the metrics (assuming more recent versions of Kiali and using Istio 1.8+).   This is the preferred way to use the metrics as the canonical fields are guaranteed to be set in a predictable way.   If you are setting both the `app.kubernetes.io/name` and `app` labels then you should know that the `canonical` values prefer the `app.kubernetes.io` setting.  The `app.kubernetes.io/version` and `version` labels act similarly.

It's not really expected to be using both labeling schemes but if you are, they should be set to the same values.  It sounds to me like you have differing values and that is the source of the problem.    I would suggest using only a single labeling scheme and then configuring Kiali for that scheme.  If for some reason you need both then setting them the same should solve the problem, otherwise I would suggest configuring Kiali to use the `app.kubernetes.io` labels.

One caveat, if you are using workload entries there may be an Istio problem setting the canonical_revision.  We're investigating that currently with Istio.

**Reply by saarwasserman** [reply]
thanks. works. I saw kiali's deployment also have both labels 'app' - 'app.kuberenetes.io/name' and 'version' - 'app.kubernetes.io/version' (of course with same values)  and service selector uses 'app'. maybe the right thing to do in kiali is same mechanism as istio's. defaults to 'app.kubernetes.io/name' and then 'app' both in labels and selectors. in a large project with many apps and 3rd party, one for sure will encounter different labeling and selectors between the two.

**Reply by jmazzitelli** [reply]
>maybe the right thing to do in kiali is same mechanism as istio's. defaults to 'app.kubernetes.io/name' and then 'app' both in labels and selectors

not trivial without causing huge performance hit. This feature [was discussed](https://github.com/kiali/kiali/issues/1458#issuecomment-558271458) a while ago. The problem is k8s APIs do not allow for querying on different label NAMES in selectors. You are forced to make separate queries for the different labels you are looking for - effectively doubling at minimum the number of queries you have to make.
